File: implementations/cpp/interval_trees.cpp
/**
 * @file interval_trees.cpp
 * @brief Reference implementation of Augmented Interval Trees with Subtree Max Augmentation.
 *
 * Implements balanced Interval Tree (augmented Treap) for dynamic interval storage,
 * point stabbing queries, single-overlap search pruning, and all-overlaps reporting.
 * Follows Arthur's Two-Layer API and differential testing against linear vector oracle.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

struct Interval {
    int64_t low = 0;
    int64_t high = 0;

    Interval() = default;
    Interval(int64_t l, int64_t h) : low(l), high(h) {
        assert(low <= high);
    }

    bool overlaps(const Interval& o) const {
        return low <= o.high && o.low <= high;
    }

    bool operator==(const Interval& o) const {
        return low == o.low && high == o.high;
    }

    bool operator<(const Interval& o) const {
        if (low != o.low) return low < o.low;
        return high < o.high;
    }
};

/**
 * @brief Augmented Balanced Interval Tree.
 * Each node stores an interval [low, high] and max_high of its subtree.
 */
class IntervalTree {
private:
    struct Node {
        Interval intv;
        int64_t max_high;
        uint32_t priority;
        Node* left = nullptr;
        Node* right = nullptr;

        Node(const Interval& i, uint32_t prio)
            : intv(i), max_high(i.high), priority(prio) {}
    };

    Node* root_ = nullptr;
    size_t size_ = 0;
    std::mt19937 rng_{42};

    static int64_t get_max_high(Node* n) {
        return n ? n->max_high : INT64_MIN;
    }

    static void update_node(Node* n) {
        if (!n) return;
        n->max_high = n->intv.high;
        if (n->left) n->max_high = std::max(n->max_high, n->left->max_high);
        if (n->right) n->max_high = std::max(n->max_high, n->right->max_high);
    }

    Node* rotate_right(Node* y) {
        Node* x = y->left;
        y->left = x->right;
        x->right = y;
        update_node(y);
        update_node(x);
        return x;
    }

    Node* rotate_left(Node* x) {
        Node* y = x->right;
        x->right = y->left;
        y->left = x;
        update_node(x);
        update_node(y);
        return y;
    }

    Node* insert_internal(Node* t, const Interval& intv, uint32_t prio, bool& inserted) {
        if (!t) {
            inserted = true;
            return new Node(intv, prio);
        }

        if (intv == t->intv) {
            inserted = false;
            return t;
        }

        if (intv < t->intv) {
            t->left = insert_internal(t->left, intv, prio, inserted);
            if (t->left->priority > t->priority) {
                t = rotate_right(t);
            }
        } else {
            t->right = insert_internal(t->right, intv, prio, inserted);
            if (t->right->priority > t->priority) {
                t = rotate_left(t);
            }
        }

        update_node(t);
        return t;
    }

    Node* erase_internal(Node* t, const Interval& intv, bool& erased) {
        if (!t) return nullptr;

        if (intv < t->intv) {
            t->left = erase_internal(t->left, intv, erased);
        } else if (t->intv < intv) {
            t->right = erase_internal(t->right, intv, erased);
        } else {
            // Found node to erase
            erased = true;
            if (!t->left && !t->right) {
                delete t;
                return nullptr;
            } else if (!t->left) {
                Node* r = t->right;
                delete t;
                return r;
            } else if (!t->right) {
                Node* l = t->left;
                delete t;
                return l;
            } else {
                // Rotate with higher priority child
                if (t->left->priority > t->right->priority) {
                    t = rotate_right(t);
                    t->right = erase_internal(t->right, intv, erased);
                } else {
                    t = rotate_left(t);
                    t->left = erase_internal(t->left, intv, erased);
                }
            }
        }

        update_node(t);
        return t;
    }

    void find_all_internal(Node* t, const Interval& q, std::vector<Interval>& out) const {
        if (!t) return;

        // 1. Search left if left subtree can contain an overlap
        if (t->left && t->left->max_high >= q.low) {
            find_all_internal(t->left, q, out);
        }

        // 2. Check current node
        if (t->intv.overlaps(q)) {
            out.push_back(t->intv);
        }

        // 3. Search right only if right subtree could have an interval with low <= q.high
        if (t->right && t->intv.low <= q.high) {
            find_all_internal(t->right, q, out);
        }
    }

    void destroy(Node* t) {
        if (!t) return;
        destroy(t->left);
        destroy(t->right);
        delete t;
    }

    bool verify_node(Node* t, const Interval* min_i, const Interval* max_i) const {
        if (!t) return true;
        if (min_i && !( *min_i < t->intv )) return false;
        if (max_i && !( t->intv < *max_i )) return false;

        int64_t expected_max = t->intv.high;
        if (t->left) expected_max = std::max(expected_max, t->left->max_high);
        if (t->right) expected_max = std::max(expected_max, t->right->max_high);
        if (t->max_high != expected_max) return false;

        return verify_node(t->left, min_i, &t->intv) && verify_node(t->right, &t->intv, max_i);
    }

public:
    IntervalTree() = default;

    ~IntervalTree() {
        destroy(root_);
    }

    IntervalTree(const IntervalTree&) = delete;
    IntervalTree& operator=(const IntervalTree&) = delete;

    size_t size() const { return size_; }
    bool empty() const { return size_ == 0; }

    // --- Layer A: Fast Core API (Preconditioned) ---

    const Interval& find_overlap(const Interval& q) const {
        const Interval* res = try_find_overlap(q);
        assert(res != nullptr);
        return *res;
    }

    void insert(int64_t low, int64_t high) {
        Interval intv(low, high);
        bool inserted = false;
        root_ = insert_internal(root_, intv, rng_(), inserted);
        if (inserted) size_++;
    }

    bool erase(int64_t low, int64_t high) {
        Interval intv(low, high);
        bool erased = false;
        root_ = erase_internal(root_, intv, erased);
        if (erased) size_--;
        return erased;
    }

    // --- Layer B: Safe Adapter API ---

    const Interval* try_find_overlap(const Interval& q) const {
        Node* curr = root_;
        while (curr) {
            if (curr->intv.overlaps(q)) {
                return &curr->intv;
            }
            if (curr->left && curr->left->max_high >= q.low) {
                curr = curr->left;
            } else {
                curr = curr->right;
            }
        }
        return nullptr;
    }

    std::vector<Interval> find_all_overlaps(const Interval& q) const {
        std::vector<Interval> out;
        find_all_internal(root_, q, out);
        return out;
    }

    std::vector<Interval> stabbing_query(int64_t point) const {
        return find_all_overlaps(Interval(point, point));
    }

    bool verify_invariants() const {
        if (!root_) return size_ == 0;
        return verify_node(root_, nullptr, nullptr);
    }
};

/**
 * @brief Naive vector oracle for differential verification.
 */
class NaiveIntervalOracle {
private:
    std::vector<Interval> intervals_;

public:
    void insert(int64_t low, int64_t high) {
        Interval intv(low, high);
        if (std::find(intervals_.begin(), intervals_.end(), intv) == intervals_.end()) {
            intervals_.push_back(intv);
        }
    }

    bool erase(int64_t low, int64_t high) {
        Interval intv(low, high);
        auto it = std::find(intervals_.begin(), intervals_.end(), intv);
        if (it != intervals_.end()) {
            intervals_.erase(it);
            return true;
        }
        return false;
    }

    bool has_overlap(const Interval& q) const {
        for (const auto& i : intervals_) {
            if (i.overlaps(q)) return true;
        }
        return false;
    }

    std::vector<Interval> find_all_overlaps(const Interval& q) const {
        std::vector<Interval> out;
        for (const auto& i : intervals_) {
            if (i.overlaps(q)) out.push_back(i);
        }
        std::sort(out.begin(), out.end());
        return out;
    }

    size_t size() const { return intervals_.size(); }
};

} // namespace dsa

int main() {
    std::cout << "Running Interval Tree verification..." << std::endl;

    dsa::IntervalTree itree;
    dsa::NaiveIntervalOracle oracle;

    // 1. Classical Textbook Interval Set (CLRS Fig 14.4)
    std::vector<std::pair<int64_t, int64_t>> textbook = {
        {16, 21}, {8, 9}, {25, 30}, {5, 8}, {15, 23},
        {17, 19}, {26, 26}, {0, 3}, {6, 10}, {19, 20}
    };

    for (const auto& p : textbook) {
        itree.insert(p.first, p.second);
        oracle.insert(p.first, p.second);
    }

    assert(itree.size() == 10);
    assert(itree.verify_invariants());

    // Query [22, 25]: Should find [15, 23]
    dsa::Interval q1(22, 25);
    const dsa::Interval* found1 = itree.try_find_overlap(q1);
    assert(found1 != nullptr && found1->overlaps(q1));

    // Stabbing query at point 7: Should hit [5, 8], [6, 10]
    auto stab7 = itree.stabbing_query(7);
    assert(stab7.size() == 2);

    // 2. Randomized Differential Fuzzing vs Naive Oracle
    std::mt19937 rng(1337);
    const int OPS = 1000;

    for (int step = 0; step < OPS; ++step) {
        int op = rng() % 4;
        int64_t l = rng() % 100;
        int64_t h = l + (rng() % 30);

        if (op == 0) {
            // Insert
            itree.insert(l, h);
            oracle.insert(l, h);
        } else if (op == 1) {
            // Erase
            bool r1 = itree.erase(l, h);
            bool r2 = oracle.erase(l, h);
            assert(r1 == r2);
        } else if (op == 2) {
            // Single overlap query
            dsa::Interval q(l, h);
            const dsa::Interval* res = itree.try_find_overlap(q);
            bool expected_overlap = oracle.has_overlap(q);
            if (expected_overlap) {
                assert(res != nullptr);
                assert(res->overlaps(q));
            } else {
                assert(res == nullptr);
            }
        } else {
            // All overlaps query
            dsa::Interval q(l, h);
            auto all_tree = itree.find_all_overlaps(q);
            auto all_oracle = oracle.find_all_overlaps(q);
            std::sort(all_tree.begin(), all_tree.end());
            assert(all_tree == all_oracle);
        }

        assert(itree.size() == oracle.size());
        assert(itree.verify_invariants());
    }

    std::cout << "[PASS] Textbook intervals and stabbing query verified." << std::endl;
    std::cout << "[PASS] 1000 differential operations matched naive oracle bit-for-bit." << std::endl;
    std::cout << "All Interval Tree assertions passed successfully!" << std::endl;
    return 0;
}

File: implementations/python/interval_trees.py
"""
Reference implementation of Augmented Interval Trees with Subtree Max Augmentation.

Implements balanced Interval Tree (augmented Treap) for dynamic interval storage,
point stabbing queries, single-overlap search pruning, and all-overlaps reporting.
Verified against a linear naive oracle.
"""

import random
import unittest
from typing import Optional, List, Tuple


class Interval:
    __slots__ = ('low', 'high')

    def __init__(self, low: int, high: int):
        assert low <= high, f"Invalid interval: low ({low}) must be <= high ({high})"
        self.low = low
        self.high = high

    def overlaps(self, other: 'Interval') -> bool:
        return self.low <= other.high and other.low <= self.high

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Interval):
            return False
        return self.low == other.low and self.high == other.high

    def __lt__(self, other: 'Interval') -> bool:
        if self.low != other.low:
            return self.low < other.low
        return self.high < other.high

    def __repr__(self) -> str:
        return f"[{self.low}, {self.high}]"


class IntervalNode:
    __slots__ = ('intv', 'max_high', 'priority', 'left', 'right')

    def __init__(self, intv: Interval, priority: int):
        self.intv = intv
        self.max_high = intv.high
        self.priority = priority
        self.left: Optional['IntervalNode'] = None
        self.right: Optional['IntervalNode'] = None


class IntervalTree:
    """Augmented Balanced Interval Tree (Treap-based) with O(log N) operations."""

    def __init__(self):
        self.root: Optional[IntervalNode] = None
        self._size = 0
        self._rng = random.Random(42)

    def __len__(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self.root is None

    @staticmethod
    def _update_node(n: Optional[IntervalNode]) -> None:
        if not n:
            return
        m = n.intv.high
        if n.left:
            m = max(m, n.left.max_high)
        if n.right:
            m = max(m, n.right.max_high)
        n.max_high = m

    def _rotate_right(self, y: IntervalNode) -> IntervalNode:
        x = y.left
        assert x is not None
        y.left = x.right
        x.right = y
        self._update_node(y)
        self._update_node(x)
        return x

    def _rotate_left(self, x: IntervalNode) -> IntervalNode:
        y = x.right
        assert y is not None
        x.right = y.left
        y.left = x
        self._update_node(x)
        self._update_node(y)
        return y

    def insert(self, low: int, high: int) -> None:
        intv = Interval(low, high)
        prio = self._rng.randint(0, 1 << 30)
        inserted = False

        def _insert(t: Optional[IntervalNode]) -> IntervalNode:
            nonlocal inserted
            if not t:
                inserted = True
                return IntervalNode(intv, prio)
            if intv == t.intv:
                return t
            if intv < t.intv:
                t.left = _insert(t.left)
                if t.left.priority > t.priority:
                    t = self._rotate_right(t)
            else:
                t.right = _insert(t.right)
                if t.right.priority > t.priority:
                    t = self._rotate_left(t)
            self._update_node(t)
            return t

        self.root = _insert(self.root)
        if inserted:
            self._size += 1

    def erase(self, low: int, high: int) -> bool:
        intv = Interval(low, high)
        erased = False

        def _erase(t: Optional[IntervalNode]) -> Optional[IntervalNode]:
            nonlocal erased
            if not t:
                return None
            if intv < t.intv:
                t.left = _erase(t.left)
            elif t.intv < intv:
                t.right = _erase(t.right)
            else:
                erased = True
                if not t.left and not t.right:
                    return None
                elif not t.left:
                    return t.right
                elif not t.right:
                    return t.left
                else:
                    if t.left.priority > t.right.priority:
                        t = self._rotate_right(t)
                        t.right = _erase(t.right)
                    else:
                        t = self._rotate_left(t)
                        t.left = _erase(t.left)
            self._update_node(t)
            return t

        self.root = _erase(self.root)
        if erased:
            self._size -= 1
        return erased

    def find_overlap(self, query: Interval) -> Optional[Interval]:
        """Find any interval overlapping with query using O(log N) pruning."""
        curr = self.root
        while curr:
            if curr.intv.overlaps(query):
                return curr.intv
            if curr.left and curr.left.max_high >= query.low:
                curr = curr.left
            else:
                curr = curr.right
        return None

    def find_all_overlaps(self, query: Interval) -> List[Interval]:
        """Find all intervals overlapping with query in O(k + log N) time."""
        out: List[Interval] = []

        def _search(t: Optional[IntervalNode]):
            if not t:
                return
            if t.left and t.left.max_high >= query.low:
                _search(t.left)
            if t.intv.overlaps(query):
                out.append(t.intv)
            if t.right and t.intv.low <= query.high:
                _search(t.right)

        _search(self.root)
        return out

    def stabbing_query(self, point: int) -> List[Interval]:
        """Find all intervals containing point."""
        return self.find_all_overlaps(Interval(point, point))

    def verify_invariants(self) -> bool:
        if not self.root:
            return self._size == 0

        def check(n: Optional[IntervalNode], min_i: Optional[Interval], max_i: Optional[Interval]) -> bool:
            if not n:
                return True
            if min_i and not (min_i < n.intv):
                return False
            if max_i and not (n.intv < max_i):
                return False

            expected_max = n.intv.high
            if n.left:
                expected_max = max(expected_max, n.left.max_high)
            if n.right:
                expected_max = max(expected_max, n.right.max_high)
            if n.max_high != expected_max:
                return False

            return check(n.left, min_i, n.intv) and check(n.right, n.intv, max_i)

        return check(self.root, None, None)


class NaiveIntervalOracle:
    def __init__(self):
        self.intervals: List[Interval] = []

    def insert(self, low: int, high: int) -> None:
        intv = Interval(low, high)
        if intv not in self.intervals:
            self.intervals.append(intv)

    def erase(self, low: int, high: int) -> bool:
        intv = Interval(low, high)
        if intv in self.intervals:
            self.intervals.remove(intv)
            return True
        return False

    def has_overlap(self, q: Interval) -> bool:
        return any(i.overlaps(q) for i in self.intervals)

    def find_all_overlaps(self, q: Interval) -> List[Interval]:
        out = [i for i in self.intervals if i.overlaps(q)]
        out.sort()
        return out

    def __len__(self) -> int:
        return len(self.intervals)


class TestIntervalTree(unittest.TestCase):
    def test_textbook_network(self):
        itree = IntervalTree()
        oracle = NaiveIntervalOracle()

        textbook = [
            (16, 21), (8, 9), (25, 30), (5, 8), (15, 23),
            (17, 19), (26, 26), (0, 3), (6, 10), (19, 20)
        ]
        for l, h in textbook:
            itree.insert(l, h)
            oracle.insert(l, h)

        self.assertEqual(len(itree), 10)
        self.assertTrue(itree.verify_invariants())

        # Overlap with [22, 25] -> should find [15, 23]
        q1 = Interval(22, 25)
        found = itree.find_overlap(q1)
        self.assertIsNotNone(found)
        self.assertTrue(found.overlaps(q1))

        # Stabbing at 7 -> should hit [5, 8], [6, 10]
        stab = itree.stabbing_query(7)
        self.assertEqual(len(stab), 2)

    def test_differential_fuzzing(self):
        itree = IntervalTree()
        oracle = NaiveIntervalOracle()

        rng = random.Random(1337)
        for _ in range(1000):
            op = rng.randint(0, 3)
            l = rng.randint(0, 100)
            h = l + rng.randint(0, 30)

            if op == 0:
                itree.insert(l, h)
                oracle.insert(l, h)
            elif op == 1:
                r1 = itree.erase(l, h)
                r2 = oracle.erase(l, h)
                self.assertEqual(r1, r2)
            elif op == 2:
                q = Interval(l, h)
                res = itree.find_overlap(q)
                has = oracle.has_overlap(q)
                if has:
                    self.assertIsNotNone(res)
                    self.assertTrue(res.overlaps(q))
                else:
                    self.assertIsNone(res)
            else:
                q = Interval(l, h)
                res_all = itree.find_all_overlaps(q)
                res_oracle = oracle.find_all_overlaps(q)
                res_all.sort()
                self.assertEqual(res_all, res_oracle)

            self.assertEqual(len(itree), len(oracle))
            self.assertTrue(itree.verify_invariants())


if __name__ == '__main__':
    unittest.main()
